### PR TITLE
Dom with javac may13

### PR DESCRIPTION
**Pull request by robstryker**

### CODE REVIEW

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	public boolean isOpen() {
+		return this.moduleSymbol.isOpen();
+	}
+
+	@Override
+	public IModuleBinding[] getRequiredModules() {
+		RequiresDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.REQUIRES).map((x) -> (RequiresDirective)x).toArray(RequiresDirective[]::new);
+		IModuleBinding[] arr2 = new IModuleBinding[arr.length];
+		for( int i = 0; i < arr.length; i++ ) {
+			arr2[i] = new JavacModuleBinding((ModuleType)arr[i].module.type, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:119]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+			arr2[i] = new JavacPackageBinding((PackageSymbol)arr[i].packge, this.resolver);
+		}
+		return arr2;
+	}
+
+	@Override
+	public String[] getExportedTo(IPackageBinding packageBinding) {
+		ExportsDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.EXPORTS).map((x) -> (ExportsDirective)x).toArray(ExportsDirective[]::new);
+		for( int i = 0; i < arr.length; i++ ) {
+			JavacPackageBinding tmp = new JavacPackageBinding((PackageSymbol)arr[i].packge, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:138]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		return new String[0];
+	}
+
+	@Override
+	public IPackageBinding[] getOpenedPackages() {
+		OpensDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.OPENS).map((x) -> (OpensDirective)x).toArray(OpensDirective[]::new);
+		IPackageBinding[] arr2 = new IPackageBinding[arr.length];
+		for( int i = 0; i < arr.length; i++ ) {
+			arr2[i] = new JavacPackageBinding((PackageSymbol)arr[i].packge, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:152]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		return new String[0];
+	}
+
+	@Override
+	public ITypeBinding[] getUses() {
+		UsesDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.USES).map((x) -> (UsesDirective)x).toArray(UsesDirective[]::new);
+		ITypeBinding[] arr2 = new ITypeBinding[arr.length];
+		for( int i = 0; i < arr.length; i++ ) {
+			arr2[i] = new JavacTypeBinding(arr[i].getService().type, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:175]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		return arr2;
+	}
+
+	@Override
+	public ITypeBinding[] getServices() {
+		ProvidesDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.PROVIDES).map((x) -> (ProvidesDirective)x).toArray(ProvidesDirective[]::new);
+		ITypeBinding[] arr2 = new ITypeBinding[arr.length];
+		for( int i = 0; i < arr.length; i++ ) {
+			arr2[i] = new JavacTypeBinding(arr[i].getService().type, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:185]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	@Override
+	public ITypeBinding[] getImplementations(ITypeBinding service) {
+		ProvidesDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.PROVIDES).map((x) -> (ProvidesDirective)x).toArray(ProvidesDirective[]::new);
+		for( int i = 0; i < arr.length; i++ ) {
+			JavacTypeBinding tmp = new JavacTypeBinding(arr[i].getService().type, this.resolver);
+			if(service.getKey().equals(tmp.getKey())) {
+				// we have our match
+				JavacTypeBinding[] ret = arr[i].getImplementations().stream().map(x -> new JavacTypeBinding((ClassType)x.type, this.resolver)).toArray(JavacTypeBinding[]::new);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:197]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacPackageBinding.java [hunk 7]
```diff
@@ -83,18 +84,25 @@ public IJavaElement getJavaElement() {
...
+			return ret;
 		} catch (JavaModelException e) {
 			// TODO Auto-generated catch block
 			e.printStackTrace();
 			return null;
 		}
 	}
+	
+	public IModuleBinding getModule() {
+		return new JavacModuleBinding(this.packageSymbol.modle, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacPackageBinding.java:104]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java [hunk 8]
```diff
@@ -463,11 +467,18 @@ public ITypeBinding[] getTypeArguments() {
...
 	@Override
 	public ITypeBinding[] getTypeBounds() {
-		Type upperBound = this.type.getUpperBound();
-		if (upperBound == null) {
-			return new ITypeBinding[0];
+		Type z1 = ((ClassType)this.type).supertype_field;
+		List<Type> z2 = ((ClassType)this.type).interfaces_field;
+		ArrayList<JavacTypeBinding> l = new ArrayList<>();
+		if( z1 != null ) {
+			l.add(new JavacTypeBinding(z1, this.resolver));
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java:474]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java [hunk 9]
```diff
@@ -463,11 +467,18 @@ public ITypeBinding[] getTypeArguments() {
...
+		}
+		if( z2 != null ) {
+			for( int i = 0; i < z2.size(); i++ ) {
+				l.add(this.resolver.canonicalize(new JavacTypeBinding(z2.get(i), this.resolver)));
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java:478]: this should be `canonicalize(new ...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java [hunk 10]
```diff
@@ -636,5 +647,13 @@ public boolean isUpperbound() {
 	public boolean isWildcardType() {
 		return this.type instanceof WildcardType;
 	}
+	
+	public IModuleBinding getModule() {
+		Symbol o = this.type.tsym.owner;
+		if( o instanceof PackageSymbol ps) {
+			return new JavacModuleBinding(ps.modle, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java:654]: this should be `canonicalize(new ...)`

**Review by mickaelistria — CHANGES_REQUESTED** [review]
@datho7561 introduced `canonicalized` that seems to fix some issues as some tests (and probably some "functional" code does use identity comparison for bindings). Please use it whenever constructing a new binding.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 11]
```diff
@@ -58,7 +59,8 @@ public JavacModuleBinding(final ModuleSymbol moduleSymbol, final ModuleType modu
 	@Override
 	public IAnnotationBinding[] getAnnotations() {
 		// TODO - don't see any way to get this?
-		return null; //new IAnnotationBinding[0];
+		List<Attribute.Compound> list = moduleSymbol.getRawAttributes();
+		return list.stream().map((x) -> new JavacAnnotationBinding(x, this.resolver, this)).toArray(JavacAnnotationBinding[]::new);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:63]: this needs also `.map(resolver::canonicalize)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacBindingResolver.java [hunk 12]
```diff
@@ -198,10 +203,20 @@ ITypeBinding resolveType(AnonymousClassDeclaration anonymousClassDecl) {
 		}
 		return null;
 	}
+	ITypeBinding resolveTypeParameter(TypeParameter typeParameter) {
+		resolve();
+		JCTree javacNode = this.converter.domToJavac.get(typeParameter);
+		if (javacNode instanceof JCTypeParameter jcClassDecl) {
+			return new JavacTypeBinding(jcClassDecl.type, this);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacBindingResolver.java:231]: This should call `resolver.canonicalize(...)`

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+		}
+		return arr2;
+	}
+
+	@Override
+	public IPackageBinding[] getExportedPackages() {
+		ExportsDirective[] arr = this.moduleSymbol.getDirectives().stream().filter((x) -> x.getKind() == DirectiveKind.EXPORTS).map((x) -> (ExportsDirective)x).toArray(ExportsDirective[]::new);
+		IPackageBinding[] arr2 = new IPackageBinding[arr.length];
+		for( int i = 0; i < arr.length; i++ ) {
+			arr2[i] = new JavacPackageBinding((PackageSymbol)arr[i].packge, this.resolver);
```
Comment (mickaelistria) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacModuleBinding.java:129]: this should be canonicalize(new ...)

**Review by mickaelistria — CHANGES_REQUESTED** [review]
Some missing canonicalize.
But everything else seems good.

**Comment by robstryker** [comment]
Merging without canocolization... we will revisit that.  It's a good point but i've had some issues and am hesitant to add that part at this moment.

**Comment by mickaelistria** [comment]
We will revisit it as part of #395

**Pull request merged**